### ISSUE TITLE
Animated volume rendering

### ISSUE THREAD

**Issue by EelcoHoogendoorn** [pygfx/pygfx#564]
Im interesting in doing a raymarched rendering of animated volume data. From looking at the texture class it only goes up to 3d. Is there something simple im missing that would allow this to be done in pygfx; or would this be a nontrivial feature request?

Ideally id like to be able to put a 5d array in gpu memory (xyzt with a number of components) and then inject a little shader code to customize how those components sampled at each voxel are mapped to rgb and accumulated though I suppose I could precompute mappings to some rgb density as well if thats more convenient.

Would it see feasible to implement something like that into pygfx, without having to fall back to pure custom shader code?

**Comment by EelcoHoogendoorn** [comment]
Thanks; yeah I was already suspecting that wgpu wouldnt have blessed us with general nd-arrays.

Its a total amount of memory that should fit into GPU mem as a whole (just a few dozen looping frames) so it probably makes most sense to put each time slice into mem as a seperate 3d texture and pre-bake the mapping to RBG into them, then bind that gpu mem resource to the volume draw call each frame.

**Comment by Korijn** [comment]
Make sure to query your hardware's maximum texture count, dimensions, bit depth, etc, so you don't hit a wall halfway through your project :)

**Comment by almarklein** [comment]
> so it probably makes most sense to put each time slice into mem as a seperate 3d texture and pre-bake the mapping to RBG into them, then bind that gpu mem resource to the volume draw call each frame.

Swapping out the texture (i.e. `geometry.grid`) will cause the shader to re-compose the pipeline. It *should* be able to re-use the wgsl shader from cache, but it'd still be relatively expensive. That said, it may be fast enough.

I think technically (from the pov of wgpu) it should be possible to swap out a binding very cheaply as long as its the same format, but we don't have a way to do that (yet). Until we have such a feature, it might be more performant to just bind to all textures from the start, and then select the correct one in the shader.

I'm not sure how the frames relate in your case, but I've done rendering of 3D data in motion (many years ago), where I had a displacement field in another texture that I used during the sampling to "move the data". The displacement field had a lower resolution, which made it feasible to simply upload a new field at each frame. So if the data allows it, a bit of creativity can also come a long way :)

**Comment by EelcoHoogendoorn** [comment]
Hmm; sounds like perhaps just merging the frames into one mega-texture and copy-pasting the volume rendering shader code to hack in a time-axis interpolation might be a fairly simple, performant and flexible approach?

**Comment by almarklein** [comment]
Yes, I think that approach might be very reasonable. Especially if the texture limits still allow you to put all volumes side-by-side (no tiling in 2 or 3 dimensions).

**Comment by pieper** [comment]
I've done something like this megatexture idea in the past (for 3D ultrasound loops) by stacking the frames on top of each other, kind of like a film strip in one texture and then just changing the texture coordinates for each timepoint.  It was very fast if all the data fits in GPU memory.
